scIB/integration.py

```diff
@@ -174,6 +174,69 @@ def runScvi(adata, batch, hvg=None):
     adata.obsm['X_emb'] = latent
 
     return adata
+def runScanvi(adata, batch, labels, hvg=None):
+    # Use non-normalized (count) data for scvi!
+    # Expects data only on HVGs
+    
+    scIB.utils.checkSanity(adata, batch, hvg)
+
+    # Check for counts data layer
+    if 'counts' not in adata.layers:
+        raise TypeError('Adata does not contain a `counts` layer in `adata.layers[`counts`]`')
+
+    from scvi.models import VAE, SCANVI
+    from scvi.inference import AlternateSemiSupervisedTrainer, SemiSupervisedTrainer
+    from sklearn.preprocessing import LabelEncoder
+    from scvi.dataset import AnnDatasetFromAnnData
+
+    # Defaults from SCVI github tutorials scanpy_pbmc3k and harmonization
+    n_epochs=np.min([round((20000/adata.n_obs)*400), 400])
+    n_latent=30
+    n_hidden=128
+    n_layers=2
+    
+    net_adata = adata.copy()
+    net_adata.X = adata.layers['counts']
+    del net_adata.layers['counts']
+    # Ensure that the raw counts are not accidentally used
+    del net_adata.raw # Note that this only works from anndata 0.7
+
+    # Define batch indices
+    le = LabelEncoder()
+    net_adata.obs['batch_indices'] = le.fit_transform(net_adata.obs[batch].values)
+    net_adata.obs['labels'] = le.fit_transform(net_adata.obs[labels].values)
+
+    net_adata_out = AnnDatasetFromAnnData(net_adata)
+    net_adata_out.labels = net_adata.obs['labels']
+
+    scanvi = SCANVI(
+        net_adata_out.nb_genes,
+        reconstruction_loss='nb',
+        n_batch=net_adata_out.n_batches,
+        n_labels=net_adata_out.n_labels,
+        n_layers=n_layers,
+        #n_latent=n_latent,
+        #n_hidden=n_hidden,
+    )
+
+    trainer = SemiSupervisedTrainer(
+        scanvi,
+        net_adata_out,
+        train_size=1.0,
+        n_epochs_classifier=10,
+        use_cuda=False,
+    )
+    
+    trainer.labelled_set = trainer.create_posterior()
```

Comment:
That seems definitely super suspect. In scANVI you need to redefine what is the labelled set and what is the unlabelled set systematically. I don't know what this line would do exactly. Does it just use all cells as labelled? 

Which cell type information do you want to give to scANVI? All the cell labels? Or only a few of them? Do you have some "unlabelled" inside of the labels field?

Usually we use only label from one dataset for example or only a couple of labels from a single dataset. It does not seem meaningful to use scANVI like this? I'm unsure. Let me know if what I say makes sense.

Reply:
Hey! Thanks for your comments!

We would want to all of the labels that we have for integration, we would not have any unlabelled cells. What we basically wanted to to is to use scANVI in a similar way to scVI but with added cell type labels.
I'm not really sure if we can use scANVI in such a way.

Reply:
The though process here is that adding more prior information, on which you condition your latent variable z will improve the latent representation. So we pass all labels to get the best possible representation. If this approach doesn't make sense to you, maybe we can also omit scANVI and only compare scVI though, as then conceptually there might not be much difference except the added layer of complexity.

Reply:
Hi Malte, 

I am going to add in this PR a piece of code to do that. The idea is that the labels should help structure the latent space better, and be some form of additional information. 

One thing that would be good to add is that usually, scANVI is used for probabilistic annotations of cells, and when it is fed the complete annotation, it won't be probabilistic in the cell types anymore. Consequently, the latent space becomes biased with respect to the fed labels. Hopefully, this model should get better clustering results than scVI for example. It might deform the latent space in other ways also. 

Also, I would insist on the fact that usually not all the labels are given. But I agree this is nice to add it as a "cell type informed" latent space and see what comes out of it.

Reply:
I will be careful to mention this in the methods section and if space permits also allude to it in the discussion. Thanks for your help, @romain-lopez!

Reply:
Great thanks!